rfcs/input-validations.md

```diff
@@ -0,0 +1,365 @@
+
+### Response
+
+The HTTP handler should always return the JSON object with `200` status code. The object should contain `is_valid` field
+whose value is a boolean and an optional `error` field with a message forwarded to client when `is_valid` is `false`.
+
+```http
+200 OK
+
+{
+    "is_valid": true
+}
+```
+or
+```http
+200 OK
+
+{
+    "is_valid": false,
+    "error": "Phone number is invalid"
+}
+```
+
+When response with status other than `200` is received, Hasura raises internal exception.
+
+## Examples
+
+1. [Single model] Check if `email` is valid when creating a `user`
+
+Consider the following mutation:
+
+```graphql
+mutation insertUser($email:String, $name: String) {
+  insert_users(objects: [{email: $email, name: $name}]) {
+    affected_rows
+    returning {
+      id
+    }
+  }
+}
+```
+
+The arguments used in `users` model i.e. `$email` and `$name` are sent to the `validate_input` handler of `users` insert permission.
+The handler can check the `email` value. If the handler returns success, then the mutation proceeds else the error from the handler is forwarded.
+
+2. [Multiple models] Check if `article` length is less than 1000 when inserting an `author` with their `articles`.
+
+```graphql
+mutation insertAuthorWithArticles($name: String, $email:String, $articles_content:[article_insert_input!]!) {
+  insert_author(objects: {name: $name, email: $email, articles: {data: $articles_content}}){
+    returning {
+      id
+      articles {
+        id
+      }
+    }
+  }
+}
+
+```
+
+The arguments used in `author` model i.e. `$name`, `$email` and `$articles_content` (relationship arguments are also considered part of model arguments)  are sent to the `validate_input` handler of `author` model.
+
+The arguments used in `article` model i.e. `$articles_content` is sent to the `validate_input` handler of `article` model.
+
+If both handlers return success, then the mutation proceeds else the error(s) from the handler(s) is forwarded.
+
+## Alternative approach
+
+The `validate_input` as a new expression in the `check` clause of `Insert` permission and can be clubbed with other check permissions.
+Find a sample configuration below.
+
+```yaml
+type: pg_create_insert_permission
+args:
+  table: article,
+  source: default,
+  role: user,
+  permission:
+    check:
+      author_id: "X-HASURA-USER-ID"
+      _validate_input:
+         type: http
+         definition:
+           handler: http://www.somedomain.com/validateArticle
+    set
+      id: "X-HASURA-USER-ID"
+    columns: ["name","author_id"]
+```
+The `check` expressions use boolean syntax and support both `AND` and `OR` operations. However, combining validation
+logic with the check expressions can make it difficult for users to understand and maintain. Users mental model may not
+be clear when working with input validation. It is easy to comprehend that validation always happens first when it is
+isolated from the `check`.
+
+
+## Updates and Deletes
```

Comment:
Let's also take this up and define separate sections for Inserts, updates and Deletes respectively